### PR TITLE
Fix etcd healthcheck for consensus failures

### LINKED ISSUE #72796: Kubernetes API Server healthcheck not failing if etcd is partitioned
**What happened**:
I'm running a K8s 1.13 cluster and a partial netsplit happened due to a defective linecard dropping only 
TCP packets above a certain sizes between two masters, but the third could still communicate with both (so the fault made the network non-transitive). etcd was run on the physical network, but Kubernetes traffic runs on a Wireguard overlay for security and routing is provided by OSPF (so internal communication was not affected since it was tunelled over UDP). Etcd experienced some temporary issues due to the connections dropping only certain packets (above ~1000 bytes), but quickly dropped one host out of the quorum, at which point it was degraded but all the remaining hosts could communicate freely. Each master node has an etcd instance and an apiserver. All apiservers are anycasted inside the overlay to balance the load and provide redundancy and the announcement is coupled to the health check, so in case any apiserver fails no traffic will be routed to it. The problem was that the apiserver did not fail its healthcheck even though all etcd write requests timed out. The non-working apiserver was still accessible and got routed its share of the anycast traffic since that was routed over the UDP overlay. Thus ~1/3 of all workloads and nodes were stuck on a non-working apiserver until I nuked the non-working apiserver.

TLDR: One etcd server started timing out all writes due to a non-transitive netsplit and the Kubernetes apiserver healthcheck did not fail, causing 1/3 of K8s clients to not have a working control plane until manual intervention.
**What you expected to happen**:
The health check should've failed (in my opinion, I'm happy to discuss if you think otherwise).
**How to reproduce it (as minimally and precisely as possible)**:
Set up 3 etcds with corresponding apiservers, introduce a split between two (shouldn't have to be a packet size limit, a full disconnect should reproduce cleanly), observe that all three report healthy even though at least one of them no longer works.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.13.0
- Cloud provider or hardware configuration: Pysical x86_64 servers
- OS (e.g. from /etc/os-release): Custom
- Kernel (e.g. `uname -a`): Custom, based on 4.19. No patches affect this functionality.
- Install tools: Custom
- Others:

### PR DESCRIPTION
**Pull request by lorenz**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Changes etcd healthcheck to be in line with the one used by etcd themselves and fixes
**Which issue(s) this PR fixes**:
Fixes  #72796

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by lorenz** [comment]
Just adding some more context: The reason the members listing didn't work is because it doesn't go through consensus. So if you have (for example) network issues etcd can't read or write but will still know the members and respond to that request making the health check succeed.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 1]
```diff
@@ -72,10 +74,12 @@ func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
 		client := clientValue.Load().(*clientv3.Client)
 		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 		defer cancel()
-		if _, err := client.Cluster.MemberList(ctx); err != nil {
```
Comment (dims) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: @lorenz etcd folks are running the client.Get on each individual endpoint i believe

See endpointsFromCluster where they run MemberList and then subsequently do the fake get.

https://github.com/etcd-io/etcd/blob/master/etcdctl/ctlv3/command/ep_command.go#L97-L121

Comment (lorenz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: Correct, but as far as I know kube-apiserver only ever connects to one etcd server, right? So it doesn't matter if the cluster is healthy, just the instance it connects to.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 2]
```diff
@@ -72,10 +74,12 @@ func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
...
-			return fmt.Errorf("error listing etcd members: %v", err)
+		// See https://github.com/etcd-io/etcd/blob/master/etcdctl/ctlv3/command/ep_command.go#L118
+		_, err := client.Get(ctx, "health")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: under what circumstances will this return unhealthy?

Comment (lorenz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: According to etcd's guarantees this will fail when the instance it is connected to fails to contact a majority of etcd servers and ask for the data. Practically this means it will succeed as long as the etcd instance it is connected to will return or accept data.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: I think this makes sense since (as I understand it) a 200 from the etcd health endpoint should establish that we have quorum (@jpbetz to confirm though, he knows way better than I).

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 3]
```diff
@@ -72,10 +74,12 @@ func newETCD3HealthCheck(c storagebackend.Config) (func() error, error) {
...
+		if err == nil || err == rpctypes.ErrPermissionDenied {
+			return nil
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: I don't think we should report healthy on `rpctypes.ErrPermissionDenied`.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: etcd 3.3+ servers: https://github.com/etcd-io/etcd/blob/release-3.3/etcdserver/api/etcdhttp/metrics.go#L94-L123
etcd 3.2 server: https://github.com/etcd-io/etcd/blob/release-3.2/etcdserver/api/etcdhttp/base.go#L82-L103
etcd 3.1 server: https://github.com/etcd-io/etcd/blob/release-3.1/etcdserver/api/v2http/client.go#L369

All ultimately make a quorum read to ensure a healthy etcd cluster before reporting healthy and will return a http 503 Service Unavailable response if not.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 3 above)
Comment (lorenz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: I think you're right, I took this from etcd's healthcheck but it doesn't make a whole lot of sense for K8s which needs API access in order to function

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 2 above)
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: I just talked this over with @jpbetz and this doesn't actually do what we initially thought it was doing. This does not actually hit the etcd health endpoint, it actually just makes a query to etcd to get the value for the key 'health'. From a readability perspective, this is really easy code to misunderstand.

However, making a read against etcd here is not a bad idea since it, as a side effect, establishes quorum. @jpbetz suggested making a read against an empty range rather than against the key 'health' since we will avoid hitting a key that may be actually be in use (and potentially have a large value).

Comment (lorenz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: The old version didn't, but since I removed the permission denied handler (K8s should have access otherwise the health check is no good in case of a permission problem) this is now an issue, you're right. Using the configured prefix is probably best.

Comment (lorenz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: I've now fixed the health check to request something under the prefix to make sure we do not generate spurious errors when running with strict ACLs.

**Comment by liggitt** [comment]
this lgtm, will let @jpbetz tag

**Comment by jpbetz** [comment]
/lgtm

Since https://github.com/etcd-io/etcd/blob/master/etcdctl/ctlv3/command/ep_command.go#L118 uses the health key, I'm fine with using it here as well for consistency.

**Comment by lorenz** [comment]
Is there anything I still need to to to get this approved?

**Comment by jpbetz** [comment]
@lorenz Looks like we're missing approval from someone in the OWNER files.

@xiang90 Would you be willing to approve?

**Comment by lorenz** [comment]
Can somebody approve this? Has been LGTM for a week.

**Review by timothysc — APPROVED** [review]
